File: organized-skills/1password-skill/tools/src/op-env-export.ts
#!/usr/bin/env bun
/**
 * op-env-export - Export environment to .env file format
 *
 * Usage:
 *   op-env-export <name> <vault> [--format <format>] > .env
 */

import { parseArgs } from "util";
import {
  log,
  ensureAuth,
  itemExists,
  getItem,
  extractVariables,
  readSecret,
} from "./utils";

const HELP = `
op-env-export - Export environment to .env file format

USAGE:
    op-env-export <name> <vault> [OPTIONS]

ARGUMENTS:
    <name>      Name/title of the environment item
    <vault>     Vault containing the item

OPTIONS:
    --format <fmt>   Output format (default: env)
                     - env: Standard KEY=value format
                     - docker: Docker-compatible with quotes
                     - op-refs: Use op:// references (for templates)
                     - json: JSON key-value object
    --prefix <str>   Add prefix to all variable names
    --help           Show this help message

EXAMPLES:
    # Export to .env file
    op-env-export my-app-dev Development > .env

    # Docker-compatible format
    op-env-export my-app Development --format docker > .env

    # Create template with op:// references
    op-env-export my-app-prod Production --format op-refs > .env.tpl

    # JSON format
    op-env-export config Shared --format json

    # Add prefix to all variables
    op-env-export azure Shared --prefix AZURE_ > .env
`;

type ExportFormat = "env" | "docker" | "op-refs" | "json";

async function main() {
  const { values, positionals } = parseArgs({
    args: Bun.argv.slice(2),
    options: {
      format: { type: "string", default: "env" },
      prefix: { type: "string", default: "" },
      help: { type: "boolean", short: "h" },
    },
    allowPositionals: true,
  });

  if (values.help) {
    console.log(HELP);
    process.exit(0);
  }

  const [name, vault] = positionals;

  if (!name || !vault) {
    log.error("Missing required arguments: name and vault");
    console.log(HELP);
    process.exit(1);
  }

  const format = values.format as ExportFormat;
  const prefix = values.prefix || "";

  if (!["env", "docker", "op-refs", "json"].includes(format)) {
    log.error(`Unknown format: ${format}`);
    log.error("Supported formats: env, docker, op-refs, json");
    process.exit(1);
  }

  await ensureAuth();

  // Check if item exists
  if (!(await itemExists(name, vault))) {
    log.error(`Item '${name}' not found in vault '${vault}'`);
    process.exit(1);
  }

  log.info(`Exporting: ${name} from ${vault} (format: ${format})`);

  // Get item with revealed values
  const item = await getItem(name, vault, true);

  if (!item) {
    log.error("Failed to retrieve item");
    process.exit(1);
  }

  // Extract variables
  const vars = extractVariables(item);

  if (vars.size === 0) {
    log.error("No variables found in environment");
    process.exit(1);
  }

  // Get revealed values for each variable
  const revealedVars = new Map<string, string>();
  for (const [key] of vars) {
    const value = await readSecret(vault, name, key) || vars.get(key) || "";
    revealedVars.set(key, value);
  }

  // Output header comment
  const timestamp = new Date().toISOString();

  switch (format) {
    case "env":
    case "docker":
      console.log(`# Generated from 1Password: ${name} (${vault})`);
      console.log(`# Date: ${timestamp}`);
      console.log();
      break;
    case "op-refs":
      console.log(`# 1Password Template: ${name} (${vault})`);
      console.log(`# Use with: op inject -i .env.tpl -o .env`);
      console.log(`# Or: op run --env-file .env.tpl -- command`);
      console.log();
      break;
  }

  // Output variables
  switch (format) {
    case "env":
      for (const [key, value] of revealedVars) {
        const escaped = value.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
        console.log(`${prefix}${key}=${escaped}`);
      }
      break;

    case "docker":
      for (const [key, value] of revealedVars) {
        const escaped = value.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
        console.log(`${prefix}${key}="${escaped}"`);
      }
      break;

    case "op-refs":
      for (const [key] of revealedVars) {
        console.log(`${prefix}${key}=op://${vault}/${name}/variables/${key}`);
      }
      break;

    case "json":
      const obj: Record<string, string> = {};
      for (const [key, value] of revealedVars) {
        obj[`${prefix}${key}`] = value;
      }
      console.log(JSON.stringify(obj, null, 2));
      break;
  }

  log.info("Export complete");
}

main().catch((err) => {
  log.error(err.message);
  process.exit(1);
});

File: organized-skills/1password-skill/tools-python/src/op_env/op_env_export.py
"""op-env-export — Export environment to .env file format (Python SDK version).

Usage:
    op-env-export <name> <vault> [--format <fmt>] [--prefix <str>] > .env
"""

from __future__ import annotations

import argparse
import asyncio
import json
import sys
from datetime import datetime, timezone

from .utils import (
    extract_variables,
    find_item_by_title,
    get_client,
    log,
    resolve_vault_id,
)

HELP_EPILOG = """\
examples:
  # Export to .env file
  uv run op-env-export my-app-dev Development > .env

  # Docker-compatible format
  uv run op-env-export my-app Development --format docker > .env

  # Create template with op:// references
  uv run op-env-export my-app-prod Production --format op-refs > .env.tpl

  # JSON format
  uv run op-env-export config Shared --format json

  # Add prefix to all variables
  uv run op-env-export azure Shared --prefix AZURE_ > .env
"""

VALID_FORMATS = ("env", "docker", "op-refs", "json")


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        prog="op-env-export",
        description="Export environment to .env file format",
        epilog=HELP_EPILOG,
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument("name", help="Name/title of the environment item")
    parser.add_argument("vault", help="Vault containing the item")
    parser.add_argument(
        "--format",
        dest="fmt",
        default="env",
        choices=VALID_FORMATS,
        help="Output format (default: env)",
    )
    parser.add_argument("--prefix", default="", help="Add prefix to all variable names")
    return parser


async def main() -> None:
    parser = build_parser()
    args = parser.parse_args()

    client = await get_client()
    vault_id = await resolve_vault_id(client, args.vault)

    # Find the item
    item = await find_item_by_title(client, vault_id, args.name)
    if item is None:
        log.error(f"Item '{args.name}' not found in vault '{args.vault}'")
        sys.exit(1)

    log.info(f"Exporting: {args.name} from {args.vault} (format: {args.fmt})")

    # Extract variable names
    variables = extract_variables(item)
    if not variables:
        log.error("No variables found in environment")
        sys.exit(1)

    # Resolve actual values using batch resolution
    refs = [f"op://{args.vault}/{args.name}/variables/{key}" for key in variables]
    try:
        resolved_values = await client.secrets.resolve_all(refs)
    except Exception:
        # Fallback: resolve individually
        resolved_values = []
        for ref in refs:
            try:
                resolved_values.append(await client.secrets.resolve(ref))
            except Exception:
                resolved_values.append(variables[list(variables.keys())[len(resolved_values)]])

    revealed: dict[str, str] = dict(zip(variables.keys(), resolved_values))

    # Output header
    timestamp = datetime.now(timezone.utc).isoformat()
    prefix = args.prefix

    if args.fmt in ("env", "docker"):
        print(f"# Generated from 1Password: {args.name} ({args.vault})")
        print(f"# Date: {timestamp}")
        print()
    elif args.fmt == "op-refs":
        print(f"# 1Password Template: {args.name} ({args.vault})")
        print("# Use with: op inject -i .env.tpl -o .env")
        print("# Or: op run --env-file .env.tpl -- command")
        print()

    # Output variables
    if args.fmt == "env":
        for key, value in revealed.items():
            escaped = value.replace("\\", "\\\\").replace('"', '\\"')
            print(f"{prefix}{key}={escaped}")
    elif args.fmt == "docker":
        for key, value in revealed.items():
            escaped = value.replace("\\", "\\\\").replace('"', '\\"')
            print(f'{prefix}{key}="{escaped}"')
    elif args.fmt == "op-refs":
        for key in revealed:
            print(f"{prefix}{key}=op://{args.vault}/{args.name}/variables/{key}")
    elif args.fmt == "json":
        obj = {f"{prefix}{key}": value for key, value in revealed.items()}
        print(json.dumps(obj, indent=2))

    log.info("Export complete")


def main_sync() -> None:
    asyncio.run(main())


if __name__ == "__main__":
    main_sync()
